src/svm/utils/matutils.c

```diff
@@ -83,141 +83,250 @@ PetscErrorCode MatMultTranspose_Biased(Mat A,Vec x,Vec y)
+
+  Level: advanced
+
+.seealso MatBiasedGetBias(), MatBiasedGetInnerMat()
+@*/
+PetscErrorCode MatBiasedCreate(Mat A,PetscReal bias,Mat *A_biased)
 {
   MPI_Comm    comm;
   PetscMPIInt comm_size,comm_rank;
 
-  Mat          Xt_biased_inner;
-  PetscInt     m,n,M,N;
+  Mat         A_biased_inner;
+  PetscInt    m,n,M,N;
+  MatCtx      *ctx = NULL;
 
-  MatBiasedCtx *ctx = NULL;
+  const char  *A_name,*A_prefix;
+  char        A_name_inner[50];
 
   PetscFunctionBegin;
-  PetscObjectGetComm((PetscObject) Xt,&comm);
+  PetscValidHeaderSpecific(A,MAT_CLASSID,1);
+  PetscValidLogicalCollectiveReal(A,bias,2);
+  PetscValidPointer(A_biased,3);
+
+  PetscObjectGetComm((PetscObject) A,&comm);
   TRY( MPI_Comm_size(comm,&comm_size) );
   TRY( MPI_Comm_rank(comm,&comm_rank) );
 
+  /* Create matrix context */
   TRY( PetscNew(&ctx) );
-  ctx->Xt   = Xt;
+  ctx->inner    = A;
   ctx->bias = bias;
 
-  TRY( MatGetLocalSize(Xt,&m,&n) );
-  TRY( MatGetSize(Xt,&M,&N) );
+  TRY( MatGetLocalSize(A,&m,&n) );
+  TRY( MatGetSize(A,&M,&N) );
 
   if (comm_rank == comm_size - 1) n += 1;
 
-  TRY( MatCreateShell(comm,m,n,M,N+1,ctx,&Xt_biased_inner) );
-  TRY( MatShellSetOperation(Xt_biased_inner,MATOP_DESTROY,(void(*)(void))MatDestroy_Biased) );
-  TRY( MatShellSetOperation(Xt_biased_inner,MATOP_MULT,(void(*)(void))MatMult_Biased) );
-  TRY( MatShellSetOperation(Xt_biased_inner,MATOP_MULT_TRANSPOSE,(void(*)(void))MatMultTranspose_Biased) );
-  TRY( MatShellSetOperation(Xt_biased_inner,MATOP_CREATE_SUBMATRIX,(void(*)(void))MatCreateSubMatrix_Biased) );
-  TRY( PetscObjectComposeFunction((PetscObject) Xt_biased_inner,"MatGetOwnershipIS_C",MatGetOwnershipIS_Biased) );
-
-  *Xt_biased = Xt_biased_inner;
+  TRY( MatCreateShell(comm,m,n,M,N+1,ctx,&A_biased_inner) );
+  /* Set shell matrix functions */
+  TRY( MatShellSetOperation(A_biased_inner,MATOP_DESTROY         ,(void(*)(void))MatDestroy_Biased) );
+  TRY( MatShellSetOperation(A_biased_inner,MATOP_MULT            ,(void(*)(void))MatMult_Biased) );
+  TRY( MatShellSetOperation(A_biased_inner,MATOP_MULT_TRANSPOSE  ,(void(*)(void))MatMultTranspose_Biased) );
+  TRY( MatShellSetOperation(A_biased_inner,MATOP_CREATE_SUBMATRIX,(void(*)(void))MatCreateSubMatrix_Biased) );
+  TRY( PetscObjectComposeFunction((PetscObject) A_biased_inner,"MatGetOwnershipIS_C",MatGetOwnershipIS_Biased) );
+
+  /* Set names of mat and inner mat */
+  TRY( PetscObjectGetName((PetscObject) A,&A_name) );
+  TRY( PetscObjectSetName((PetscObject) A_biased_inner,A_name) );
+  /* Set prefixes of mat and inner mat */
+  TRY( PetscObjectGetOptionsPrefix((PetscObject) A,&A_prefix) );
+  TRY( PetscObjectSetOptionsPrefix((PetscObject) A_biased_inner,A_prefix) );
+  TRY( PetscObjectAppendOptionsPrefix((PetscObject) A,"inner_") );
+  /* Set name of inner mat */
+  TRY( PetscStrcpy(A_name_inner,A_name) );
+  TRY( PetscStrcat(A_name_inner,"_inner") );
+  TRY( PetscObjectSetName((PetscObject) A,A_name_inner) );
```

Comment:
```suggestion
  TRY( PetscObjectSetName((PetscObject) A_biased_inner,A_name_inner) );
```

Reply:
Please, read code carefully. It's not a bug. A is original matrix, A_biased_inner is biased matrix created and returned. This part is right.

Reply:
Sorry, you are right. Still, you should not change the name of the input matrix. Consider naming the new matrix e.g., A_name+"_biased".

Reply:
Accepted.